enso/resample/__init__.py

```diff
@@ -1,46 +1,124 @@
+import itertools
+from enso.config import MODE
 
 import numpy as np
 
+from enso.registry import Registry, ModeKeys
 
-def oversample(X, y, max_ratio=50):
-    """
-    Ensure each class occurs with approximately even frequency in the training set by
-    duplicating examples from relatively rare classes.
-
-    :param X: `np.ndarray` of input features
-    :param y: `np.ndarray` of corresponding targets
-    :param max_ratio: input examples should be duplicated no more than this amount
-    """
-    X, y = np.asarray(X), np.asarray(y)
-    class_counts = Counter(y)
-    max_count = max(class_counts.values())
-    desired_counts = {
-        class_name: min(
-            max_count, max_ratio * class_count)
-        for class_name, class_count in class_counts.items()
-    }
-
-    idxs_by_y = defaultdict(list)
-    for i, element in enumerate(y):
-        idxs_by_y[element].append(i)
-
-    idx_sample = []
-    for class_name in class_counts:
-        sample_idxs = np.random.choice(idxs_by_y[class_name], desired_counts[class_name], replace=True).tolist()
-        idx_sample.extend(sample_idxs)
-
-    random.shuffle(idx_sample)
-
-    return X[idx_sample].tolist(), y[idx_sample].tolist()
-
-
-def resample(resample_type, train_data, train_labels):
-    if resample_type.lower() == 'none':
-        return train_data, train_labels
-    elif resample_type == 'RandomOverSampler':
-        train_data, train_labels = oversample(train_data, train_labels)
-        return train_data, train_labels
-    else:
-        raise Exception("Invalid resample_type: {}".format(resample_type))
+
+class Resampler(metaclass=ABCMeta):
+
+    @staticmethod
+    @abstractmethod
+    def resample(X, y, max_ratio=50):
+        """ """
+
+
+@Registry.register_resampler(ModeKeys.CLASSIFY)
+class RandomOverSampler(Resampler):
+
+    @staticmethod
+    def resample(X, y, max_ratio=50):
+        X, y = np.asarray(X), np.asarray(y)
+        class_counts = Counter(y)
+        max_count = max(class_counts.values())
+        desired_counts = {
+            class_name: min(
+                max_count, max_ratio * class_count)
+            for class_name, class_count in class_counts.items()
+        }
+
+        idxs_by_y = defaultdict(list)
+        for i, element in enumerate(y):
+            idxs_by_y[element].append(i)
+
+        idx_sample = []
+        for class_name in class_counts:
+            sample_idxs = np.random.choice(idxs_by_y[class_name], desired_counts[class_name], replace=True).tolist()
+            idx_sample.extend(sample_idxs)
+
+        random.shuffle(idx_sample)
+
+        return X[idx_sample].tolist(), y[idx_sample].tolist()
+
+
+@Registry.register_resampler(ModeKeys.ANY)
+class NoneOverSampler(Resampler):
+
+    @staticmethod
+    def resample(X, y, max_ratio=50):
+        return X, y
+
+
+@Registry.register_resampler(ModeKeys.SEQUENCE)
+class SequenceOverSampler(Resampler):
+
+    @staticmethod
+    def resample(X, y, max_ratio=50):
```

Comment:
Might need to talk over this one with you in person.  Basic logic checks out.  My only concerns are around whether or not this is beneficial because it seems like it may duplicate particular high value examples (i.e. with many rare classes) many times in order to meet the requirements? This is maybe desirable but just not 100% sure.  Empirically you've shown it works so that's good enough for me but just want to make sure I'm on the same page.

Reply:
I'm not super sure that this will be clear from the code, but the flow is as follows:

- Shuffle the dataset to mean that the order of consumption is random.
- Select as many unique entries as will fit with the constraints.
- Add these once.
- Reshuffle the dataset and re-draw samples that conform to the constraints.
- Iff the selected items are exactly the same as in the last selection, it puts as many of these in as possible. This makes the assumption that if the items were shuffled and these are the only ones that fit, they are the only ones that will fit. This is not exact, sometimes this will happen purely by chance. If the assumption was correct however, this algorithm would be equivalent to random uniform oversampling sampling.

It would be possible to see if the combinations are REALLY the unique solution by removing any samples that individually violate the constraints and then trying to find an exception, but the complexity of this is not particularly nice in the worst case (I think exponential in the number of non-violating samples).